### PR TITLE
Template integration with https://github.com/xordataexchange/crypt provided values

### PR DESCRIPTION
**Pull request by philipsoutham**
Template integration with https://github.com/xordataexchange/crypt provided values

New cli option `-secret-keyring` -> path to armored secret keyring

| template function | crypt enabled function |
| --- | --- |
| `get` | `cget` |
| `gets` | `cgets` |
| `getv` | `cgetv` |
| `getvs` | `cgetvs` |

Why? Because I wanted to store SSL certs along with the nginx config values.

Also, sorry; I noticed my godeps got attached to the PR.

Closes #210 #342

### CODE REVIEW

**Comment by philipsoutham** [comment]
Lame, For whatever reason the order in 

```
{{range cgetvs "/crypt-test/*"}}
val: {{.}}
{{end}}
```

is not deterministic causing the travis test to bomb.

**Comment by kelseyhightower** [comment]
@philipsoutham Wow, this is pretty awesome. It going to take some time to review this.

**Comment by bketelsen** [comment]
so much awesome!

File: config.go [hunk 1]
```diff
@@ -45,21 +46,23 @@ var (
...
+	Interval      int      `toml:"interval"`
+	SecretKeyring string   `toml:"secret_keyring"`
+	Noop          bool     `toml:"noop"`
+	Prefix        string   `toml:"prefix"`
+	Quiet         bool     `toml:"quiet"`
+	SRVDomain     string   `toml:"srv_domain"`
+	Scheme        string   `toml:"scheme"`
+	Verbose       bool     `toml:"verbose"`
+	Watch         bool     `toml:"watch"`
+	KeyStore      []byte
```
Comment (bacongobbler) [inline config.go]: small nit, but can we change this to `PGPPrivateKey` or something that's a little more verbose? In my mind, `KeyStore` implies a storage container for managing keys, whereas this is just a reference to a PGP private key.

**Comment by bacongobbler** [comment]
Haven't tested this but from initial impressions this get my :+1: for a clean implementation.

Would you kindly add documentation around how users would go about setting this up/using the functions supplied in this PR? Thanks! :heart:

File: config.go (hunk 1 above)
Comment (philipsoutham) [inline config.go]: @bacongobbler sure thing, I was just using the taxonomy from https://github.com/xordataexchange/crypt.

**Comment by philipsoutham** [comment]
@bacongobbler I'll try to write something up to be included in the docs section to be included of the PR.

File: docs/data_encryption.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
+Encrypting your data
```
Comment (philipsoutham) [inline docs/data_encryption.md]: @bacongobbler Take a look at this documentation and let me know if you think it is adequate.

**Comment by philipsoutham** [comment]
Docs added.

File: docs/data_encryption.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+- [PGP Keypair](#local-testing-keypair)
+  - [Sample Keypair](#local-testing-keypair-sample)
+  - [Generating Your Own Keypair](#local-testing-keypair-generating)
+- [Encrypting & Storing Your Data](#local-testing-saving)
+  - [With GPG](#local-testing-saving-gpg)
+  - [With crypt](#local-testing-saving-crypt)
+- [Verifying Encrypted Data](#local-testing-verify)
+- [Putting It All Together, Using With Confd](#local-testing-confd)
+
+<a name="local-testing-keypair"/>
```
Comment (bacongobbler) [inline docs/data_encryption.md]: What's the rationale behind these internal links? Github automatically generates links like [this](https://github.com/philipsoutham/confd/blob/crypt/docs/data_encryption.md#creating-pgp-keys) on hosted markdown.

Comment (bacongobbler) [inline docs/data_encryption.md]: Ah, I see the Table of contents above. Let's remove it, as no other documentation in confd has this.

**Comment by bacongobbler** [comment]
I see that https://github.com/xordataexchange/crypt only has support for etcd and consul. What about our other backends like redis, env and eventually [the filesystem](#201)? How is someone able to encrypt data on the other backends if `crypt` doesn't have support for those backends?

Just wondering if we should only enable this for etcd and consul

**Comment by bketelsen** [comment]
PR's accepted on Crypt if you want to add support to other backends.  It's a really small interface to implement.

**Comment by philipsoutham** [comment]
The cli tool that is included as a convenience in `crypt` only supports the consul and etcd backends; the library piece that actually encrypts the data is backend agnostic. You can use the following methods included in the `github.com/xordataexchange/crypt/encoding/secconf` library to [encode](https://github.com/philipsoutham/confd/blob/8d9df2e86581471f6dc1c7aaa647de67557d93b3/resource/template/template_test.go#L241) and [decode](https://github.com/philipsoutham/confd/blob/8d9df2e86581471f6dc1c7aaa647de67557d93b3/resource/template/resource.go#L100) your data using any backed you want (that is what [we](http://itds.jpl.nasa.gov/) are doing).

File: docs/data_encryption.md (hunk 3 above)
Comment (philipsoutham) [inline docs/data_encryption.md]: TOC removed in current PR

**Comment by philipsoutham** [comment]
Rebased master.

**Comment by popsikle** [comment]
+1 on this ;)

**Comment by indiv0** [comment]
Any updates on this?

**Comment by kelseyhightower** [comment]
I'm reviewing this, we will need to rebase, but I think there are some patches floating around.

**Comment by jamiehannaford** [comment]
Can't wait for this! :rocket:

**Comment by davidquarles** [comment]
@philipsoutham This is so awesome.  Do you have time to rebase?

**Comment by sruon** [comment]
+1

**Comment by kelseyhightower** [comment]
@philipsoutham After a bit of clean up to the general confd repo I think we are ready to review and merge this. If you are still interested can you rebase this?

**Pull request merged**